File: android/src/main/kotlin/com/inappstory/inappstory_plugin/adaptors/IASOnboardingsAdaptor.kt
package com.inappstory.inappstory_plugin.adaptors

import IASOnboardingsHostApi
import com.inappstory.inappstory_plugin.callbacks.OnboardingLoadCallbackAdaptor
import com.inappstory.sdk.AppearanceManager
import com.inappstory.sdk.CancellationToken
import com.inappstory.sdk.CancellationTokenCancelResult
import com.inappstory.sdk.InAppStoryManager
import io.flutter.embedding.engine.plugins.FlutterPlugin

class IASOnboardingsAdaptor(
    flutterPluginBinding: FlutterPlugin.FlutterPluginBinding,
    private val appearanceManager: AppearanceManager,
    private val activityHolder: ActivityHolder,
) : IASOnboardingsHostApi {
    private val tokenMap = mutableMapOf<String, CancellationToken>()

    init {
        IASOnboardingsHostApi.setUp(flutterPluginBinding.binaryMessenger, this)

        InAppStoryManager.getInstance()
            ?.setOnboardingLoadCallback(OnboardingLoadCallbackAdaptor(flutterPluginBinding))
    }

    override fun show(limit: Long, feed: String, token: String, tags: List<String>) {
        val cancellationToken = InAppStoryManager.getInstance()?.showOnboardingStories(
            limit.toInt(),
            feed,
            tags,
            activityHolder.activity,
            appearanceManager,
        )
        if (cancellationToken != null) {
            tokenMap[token] = cancellationToken
        }
    }

    override fun cancelByToken(token: String): Boolean {
        if (tokenMap.containsKey(token)) {
            val result = tokenMap[token]?.cancel()
            tokenMap.remove(token)
            return result == CancellationTokenCancelResult.SUCCESS
        }
        return false
    }
}

File: ios/inappstory_plugin/Sources/inappstory_plugin/Adaptors/IASOnboardingsAdaptor.swift
import Flutter
import Foundation
@_spi(QAApp) import InAppStorySDK
@_spi(IAS_API) import InAppStorySDK

class IASOnboardingsAdaptor: IASOnboardingsHostApi {
    init(
        binaryMessenger: FlutterBinaryMessenger,
        onboardingsAPI: OnboardingsAPI
    ) {
        self.binaryMessenger = binaryMessenger

        self.onboardingsAPI = onboardingsAPI

        self.onboardingLoadCallbackFlutterApi =
            OnboardingLoadCallbackFlutterApi(binaryMessenger: binaryMessenger)

        IASOnboardingsHostApiSetup.setUp(
            binaryMessenger: binaryMessenger,
            api: self
        )
    }

    private var binaryMessenger: FlutterBinaryMessenger

    private var onboardingsAPI: OnboardingsAPI

    private var onboardingLoadCallbackFlutterApi:
        OnboardingLoadCallbackFlutterApi

    private var tokenMap: [String: InAppStorySDK.CancellationToken] = [:]

    func show(limit: Int64, feed: String, token: String, tags: [String]) throws
    {
        func complete(show: Bool) {
            if show {
                onboardingLoadCallbackFlutterApi.onboardingLoadSuccess(
                    count: limit,
                    feed: feed
                ) { _ in }
            } else {
                onboardingLoadCallbackFlutterApi.onboardingLoadError(
                    feed: feed,
                    reason: "Onboarding load error"
                ) { _ in }
            }
        }

        let cancellationToken = onboardingsAPI.showOnboarding(
            feed: feed,
            limit: Int(limit),
            with: tags,
            with: InAppStory.shared.panelSettings,
            complete: complete
        )
        tokenMap[token] = cancellationToken
    }

    func cancelByToken(token: String) throws -> Bool {
        if tokenMap[token] != nil {
            let result = tokenMap[token]!.cancel()
            tokenMap.removeValue(forKey: token)
            return result
        }
        return false
    }
}
